graphql2/graphqlapp/alert.go

```diff
@@ -177,6 +179,95 @@ func (q *Query) mergeFavorites(ctx context.Context, svcs []string) ([]string, er
 	return svcs, nil
 }
 
+// splitRangeByDuration splits the timeframe given between since and until by the duration provided.
+// Each segment is then transformed into an AlertDataPoint based on the given alerts.
+// The given alerts are required to be sorted by their CreatedAt field.
+func splitRangeByDuration(since, until time.Time, dur timeutil.ISODuration, alerts []alert.Alert) (result []graphql2.AlertDataPoint) {
+	// serialized as empty array
+	result = []graphql2.AlertDataPoint{}
+
+	if since.After(until) {
+		return result
+	}
+
+	// fill in timestamps
+	t1 := since
+	for t1.Before(until) {
+		result = append(result, graphql2.AlertDataPoint{
+			Timestamp: t1,
+		})
+		t1 = t1.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart)
+	}
+
+	if len(result) == 0 {
+		return result
+	}
+
+	i := 0
+	// ignore alerts created before since
+	for i < len(alerts) && alerts[i].CreatedAt.Before(since) {
+		i++
+	}
+
+	for ptindex := range result {
+		upperBound := until
+		if ptindex != len(result)-1 {
+			upperBound = result[ptindex+1].Timestamp
+		}
+
+		for i < len(alerts) && alerts[i].CreatedAt.Before(upperBound) {
+			result[ptindex].AlertCount++
+			i++
+		}
+	}
+
+	return result
+}
+
+func (q *Query) AlertMetrics(ctx context.Context, opts graphql2.AlertMetricsOptions) (result []graphql2.AlertDataPoint, err error) {
+	validTimeFrame, err := time.ParseDuration("1200h")
+	if err != nil {
+		return nil, err
+	}
+	err = validate.Many(
+		validate.Range("ServiceIDs", len(opts.FilterByServiceID), 1, 1),
+		validate.TimeFrame("Timeframe", opts.Since, opts.Until, validTimeFrame),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	alerts, err := q.AlertStore.Search(ctx, &alert.SearchOptions{
+		Before:    opts.Until,
+		NotBefore: opts.Since,
+		ServiceFilter: alert.IDFilter{
+			IDs:   opts.FilterByServiceID,
+			Valid: true,
+		},
+		Limit: 100,
+		Sort:  alert.SortModeDateIDReverse,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	period, err := timeutil.ParseISODuration(*opts.Period)
```

Comment:
We should validate length

```go
err := validate.ISODuration("period", period, ISODuration{TimePart: 24*time.Hour}, ISODuration{Months: 1})
if err != nil {
	return nil, err
}
```

Reply:
We could probably do ISODuration validation by UTC comparison

```go
func (dur ISODuration) Duration() time.Duration {
  t := time.Time{}.UTC()
  return t.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart).Sub(t)
}
```

or at least use that in `GT` or `Equal` comparison methods.